book/src/design/circuit/gadgets/sinsemilla/note-commit.md

```diff
@@ -0,0 +1,374 @@
+
+1. $d_0 = 1 \implies 0 \leq b_3 + 2^{4} \cdot c < 2^{126}.$
+
+   To check that $0 \leq b_3 + 2^{4} \cdot c < 2^{126},$ we use two constraints:
+
+    a) $0 \leq b_3 + 2^{4} \cdot c < 2^{140}.$ $b_3$ is already constrained individually
+       to be a $4$-bit value. $z_{c,13}$ is the index-13 running sum output by
+       $\SinsemillaHash(c).$ By constraining $$d_0 \cdot z_{c,13} = 0,$$ we constrain
+       $b_3 + 2^4 \cdot c < 2^{134} < 2^{140}.$
+
+    b) $0 \leq b_3 + 2^{4} \cdot c + 2^{140} - t_\mathbb{P} < 2^{140}$. To check this, we
+       decompose ${b_3}c' = b_3 + 2^{4} \cdot c + 2^{140} - t_\mathbb{P}$ into fourteen
+       10-bit words (little-endian) using a running sum $z_{{b_3}c'}$, looking up each
+       word in a $10$-bit lookup table. We then enforce in the custom gate that
+       $$d_0 \cdot z_{{b_3}c',14} = 0.$$
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+3 & q_{\NoteCommit,2} \cdot d_0 \cdot z_{c,13} = 0 \\\hline
+2 & q_{\NoteCommit,1} \cdot (b_3 + c \cdot 2^4 + 2^{140} - t_\mathbb{P} - {b_3}c') = 0 \\\hline
+3 & q_{\NoteCommit,2} \cdot d_0 \cdot z_{{b_3}c',14} = 0 \\\hline
+\end{array}
+$$
+
+### $\rho$ with $g_0 = 1 \implies \rho \geq 2^{254}$
+
+In these cases, we check that $\rho_{0..=253} < t_\mathbb{P} < 2^{126}$:
+
+1. $g_0 = 1 \implies 0 \leq e_1 + 2^{4} \cdot f < 2^{126}.$
+
+   To check that $0 \leq e_1 + 2^{4} \cdot f < 2^{126},$ we use two constraints:
+
+    a) $0 \leq e_1 + 2^{4} \cdot f < 2^{140}.$ $e_1$ is already constrained individually
+       to be a $4$-bit value. $z_{f,13}$ is the index-13 running sum output by
+       $\SinsemillaHash(f).$ By constraining $$g_0 \cdot z_{f,13} = 0,$$ we constrain
+       $e_1 + 2^4 \cdot f < 2^{134} < 2^{140}.$
+
+    b) $0 \leq e_1 + 2^{4} \cdot f + 2^{140} - t_\mathbb{P} < 2^{140}$. To check this, we
+       decompose ${e_1}f' = e_1 + 2^{4} \cdot f + 2^{140} - t_\mathbb{P}$ into fourteen
+       10-bit words (little-endian) using a running sum $z_{{e_1}f'}$, looking up each
+       word in a $10$-bit lookup table. We then enforce in the custom gate that
+       $$g_0 \cdot z_{{e_1}f',14} = 0.$$
+
+$$
+\begin{array}{|c|l|}
+\hline
+\text{Degree} & \text{Constraint} \\\hline
+3 & q_{\NoteCommit,2} \cdot g_0 \cdot z_{f,13} = 0 \\\hline
+2 & q_{\NoteCommit,1} \cdot (e_1 + f \cdot 2^4 + 2^{140} - t_\mathbb{P} - {e_1}f') = 0 \\\hline
+3 & q_{\NoteCommit,2} \cdot g_0 \cdot z_{{e_1}f',14} = 0 \\\hline
+\end{array}
+$$
+
+### $\psi$ with $h_1 = 1 \implies \psi \geq 2^{254}$
+
+In these cases, we check that $\psi_{0..=253} < t_\mathbb{P} < 2^{126}$:
+
+1. $h_1 = 1 \implies h_0 = 0.$
+
+   Since $h_1 = 1 \implies \psi_{0..=253} < 2^{126},$ we know that $\psi_{126..=253} = 0,$
+   and in particular $h_0 := \psi_{249..=253} = 0.$
+
+2. $h_1 = 1 \implies 0 \leq g_1 + 2^{9} \cdot g_2 < 2^{126}.$
+
+   To check that $0 \leq g_1 + 2^{9} \cdot g_2 < 2^{126},$ we use two constraints:
```

Comment:
```suggestion
In these cases, we check that $\psi_{0..=253} < t_\mathbb{P}$:

1. $h_1 = 1 \implies h_0 = 0.$

   Since $h_1 = 1 \implies \psi_{0..=253} < t_\mathbb{P} < 2^{126},$ we know that $\psi_{126..=253} = 0,$
   and in particular $h_0 := \psi_{249..=253} = 0.$

2. $h_1 = 1 \implies 0 \leq g_1 + 2^{9} \cdot g_2 < t_\mathbb{P}.$

   To check that $0 \leq g_1 + 2^{9} \cdot g_2 < t_\mathbb{P},$ we use two constraints:
```